substrate/frame/migrations/src/mock.rs

```diff
@@ -0,0 +1,297 @@
+// This file is part of Substrate.
+
+// Copyright (C) Parity Technologies (UK) Ltd.
+// SPDX-License-Identifier: Apache-2.0
+
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// 	http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#![cfg(test)]
+
+use crate::{Event, Historic, WeightMeter};
+
+use codec::{Decode, Encode};
+use core::cell::RefCell;
+use frame_support::{
+	derive_impl,
+	migrations::*,
+	traits::{OnFinalize, OnInitialize},
+	weights::Weight,
+};
+use frame_system::EventRecord;
+use sp_core::{ConstU32, H256};
+use sp_runtime::BoundedVec;
+
+type Block = frame_system::mocking::MockBlock<Test>;
+
+// Configure a mock runtime to test the pallet.
+frame_support::construct_runtime!(
+	pub enum Test {
+		System: frame_system,
+		Migrations: crate,
+	}
+);
+
+#[derive_impl(frame_system::config_preludes::TestDefaultConfig as frame_system::DefaultConfig)]
+impl frame_system::Config for Test {
+	type Block = Block;
+	type BaseCallFilter = frame_support::traits::Everything;
+	type RuntimeOrigin = RuntimeOrigin;
+	type RuntimeCall = RuntimeCall;
+	type RuntimeEvent = RuntimeEvent;
+	type PalletInfo = PalletInfo;
+	type OnSetCode = ();
+}
+
+/// An opaque identifier of a migration.
+pub type MockedIdentifier = BoundedVec<u8, ConstU32<256>>;
+
+/// How a [`MockedMigration`] should behave.
+#[derive(Debug, Clone, Copy, Encode)]
+#[allow(dead_code)]
+pub enum MockedMigrationKind {
+	/// Succeed after its number of steps elapsed.
+	SucceedAfter,
+	/// Fail after its number of steps elapsed.
+	FailAfter,
+	/// Never terminate.
+	TimeoutAfter,
+	/// Cause an [`InsufficientWeight`] error after its number of steps elapsed.
+	HightWeightAfter(Weight),
+}
+use MockedMigrationKind::*; // C style
+
+impl From<u8> for MockedMigrationKind {
+	fn from(v: u8) -> Self {
+		match v {
+			0 => SucceedAfter,
+			1 => FailAfter,
+			2 => TimeoutAfter,
+			3 => HightWeightAfter(Weight::MAX),
+			_ => unreachable!(),
+		}
+	}
+}
+
+impl Into<u8> for MockedMigrationKind {
+	fn into(self) -> u8 {
+		match self {
+			SucceedAfter => 0,
+			FailAfter => 1,
+			TimeoutAfter => 2,
+			HightWeightAfter(_) => 3,
+		}
+	}
+}
```

Comment:
don't we get `Into` for free by implementing `From`?

Reply:
Generally yes, but in this case these are not complementing. `From<u8>` provides `Into<MockedMigrationKind>`, not `Into<u8>` 😛